File: Programmers/Python3/Level 2/쿼드압축 후 개수 세기/solution.py
# https://school.programmers.co.kr/learn/courses/30/lessons/68936

def solution(arr):
    answer = [0, 0]
    check_same(arr, 0, 0, len(arr), answer)
    return answer


def check_same(arr, start_x, start_y, size, answer):
    check_num = arr[start_y][start_x]
    for y in range(start_y, start_y + size):
        for x in range(start_x, start_x + size):
            if arr[y][x] != check_num:  # 다른 수가 나왔다면 4개로 나누고 재귀탐색
                new_size = size // 2
                check_same(arr, start_x, start_y, new_size, answer)
                check_same(arr, start_x + new_size, start_y, new_size, answer)
                check_same(arr, start_x, start_y + new_size, new_size, answer)
                check_same(arr, start_x + new_size, start_y + new_size, new_size, answer)
                return
    # 사각형의 모든 숫자가 같다면 해당 숫자의 개수 증가
    answer[check_num] += 1

File: Programmers/Javascript/Level 2/쿼드압축 후 개수 세기/solution.js
// https://school.programmers.co.kr/learn/courses/30/lessons/68936?language=javascript

function solution(arr) {
  const answer = [0, 0];
  check(arr, 0, 0, arr.length, answer);
  return answer;
}

/**
 *
 * @param {} arr
 * @param {*} startX
 * @param {*} startY
 * @param {*} size
 * @param {*} answer
 * @returns
 */
function check(arr, startX, startY, size, answer) {
  const checkNumber = arr[startY][startX];

  for (let y = startY; y < startY + size; y++) {
    for (let x = startX; x < startX + size; x++) {
      if (arr[y][x] !== checkNumber) {
        const halfSize = size / 2;

        check(arr, startX, startY, halfSize, answer);
        check(arr, startX, startY + halfSize, halfSize, answer);
        check(arr, startX + halfSize, startY, halfSize, answer);
        check(arr, startX + halfSize, startY + halfSize, halfSize, answer);
        return;
      }
    }
  }

  answer[checkNumber] += 1;
}
